jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java

```diff
@@ -385,6 +385,13 @@ private boolean parseRequestBuffer()
 
     private boolean upgrade()
     {
+        // If we are fill interested, then a read is pending and we must abort
+        if (isFillInterested())
+        {
+            LOG.warn("Pending read in onCompleted {} {}", this, getEndPoint());
+            abort(new IllegalStateException());
+        }
+
```

Comment:
This is a strange unrelated side effect of `upgrade()` method.     Should be done in `onComplete()` prior to calling `upgrade()`.

Reply:
Ok, done.